feat: access token & enhanced endpoints (/login, /userinfo, /logout)

shawnhankim opened the pull request:
**Issue Item:**
  - https://github.com/nginxinc/nginx-openid-connect/pull/62

**Description:**

**1. access token**

  - Enhance the NJS Code to capture the `access_token` sent by the IdP.
  - Store the `access_token` in the k/v store as same as we store `id_token` and `refresh_token`

**2. new endpoints**

  - Add `/userinfo` endpoint: 
    - Add a map variable of `$oidc_userinfo_endpoint` as same as authz and token endpoints here (`openid_connect_configuration.conf`) . 
    - Expose `/userinfo` endpoint here(`openid_connect.server_conf`) in a location block of NGINX Plus to interact with IdP's `userinfo_endpoint` which is defined in the endpoint of`well-known/openid-configuration`. 
    - The nginx location block should proxy to the IdP’s `userinfo_endpoint` by adding `access_token` as a bearer token. 
      ```
      Authorization : Bearer <access_token>
      ```
    - The response coming from IdP should be returned back to the caller as it is.

  - Expose `/login` endpoint:
    - Expose the `/login` endpoint as a location block here (`openid_connect.server_conf`) 
    - Proxy it to the IdP's `authorization_endpoint` configured in the map variable of `$oidc_authz_endpoint` in (`openid_connect_configuration.conf`). 
    - This would outsource the login function to IdP as its configured.

  - Enhance `/logout` endpoint:
    - Add a map variable of `$oidc_end_session_endpoint` as same as authz and token endpoints here (`openid_connect_configuration.conf`) .
    - Proxy it to the IdP's `end_session_endpoint` to finish the session by IdP.

  - Enhance `/_logout` endpoint:
    - Enhance `/_logout` endpoint which is a callback from IdP as a location block here (`openid_connect.server_conf`) to handle the following sequences.
      - 1. Redirected by IdP when IdP successfully finished the session.
      - 2. NGINX Plus: Clear session cookies.
      - 3. NGINX Plus: Redirect to either the landing page or the custom logout page by calling

**3. add endpoints in `configure.sh`**

  - IdP's userinfo endpoint
  - IdP's end session endpoint

**Compatibility:**
- This PR does not block the existing customers as it just adds endpoints and features.

shawnhankim commented:
@route443 : Per our discussions,
- I have consolidated 3 PRs into 1 PR with 1 commit here. 
- I have separated the PR for doc. 
- I am going to separately provide a demo files which can be merged into a feature branch like `demo` once I have access to the branch, or if you create the branch. Until then, you can test this PR using [this repo](https://github.com/nginx-openid-connect/nginx-oidc-examples/tree/main/001-oidc-local-test).

route443 commented on frontend.conf:
```diff
@@ -1,3 +1,16 @@
+# -----------------------------------------------------------------------------#
```
In general, the current style does not imply the use of any ascii-frames. I'm not sure if this improves the readability of the config files, so I would remove it.

route443 commented on frontend.conf:
```diff
@@ -33,4 +46,106 @@ server {
 
+# -----------------------------------------------------------------------------#
+#                                                                              #
+# 2. Advanced Example: Landing page, login/logout button to handle OIDC kflow  #
```
I think it's better to have one example with commented out directives than to have multiple examples which can lead to additional complexity or potential security risks. More details below.

shawnhankim replied:
Great comment and idea to have one example!

route443 commented on frontend.conf:
```diff
@@ -33,4 +46,106 @@ server {
+#  - Landing page shows 'login' button                                         #
+#  - 'login' button calls `/login` endpoint to start OIDC flow by validating 
+#    'id_token' w/ IdP's JWK.  #
+#  - Landing page calls `/userinfo` to show user info using 'access_token`.    #
```
The Userinfo Claims are needed by the Client and is not meant to be provided to the User Agent. Why would we disclose this information to the end user?

route443 commented on frontend.conf:
```diff
@@ -33,4 +46,106 @@ server {
+}
+
+#
+# Upstream sample for proxing to the backend API server.
```
The previous remark is true for this example as well.

route443 commented on frontend.conf:
```diff
@@ -33,4 +46,106 @@ server {
+    #    used for API authorization.
+    #    + Most of IdP generate an access token for API authorization of IdP's
+    #      endpoints (like /userinfo) as well as customer's endpoints.
+    #    + But Azure AD generate two types of access token for API authorization
```
This description is also vague and technically incomplete, which can cause a lot of misunderstanding. The fact that Azure identity platform adds a nonce to the header after it has signed the token does not make it somehow inferior, it means that we should not verify this token, it is not for us, it is for Token Introspection endpoint / Graph API / etc. Again, this is a topic for a separate deployment article.

shawnhankim replied:
Sure thing, I agree with you. Thanks for your opinion!

route443 commented on openid_connect.server_conf:
```diff
@@ -66,6 +69,70 @@
 
+    #
+    # User information endpoint for the following purposes:
+    # - Browser to periodically check if your are signed-in based on status code.
```
How can the browser be forced to do this? Using custom JS code? What happens if we get a 401 Unauthorized, what does that mean? Access Token not valid? Whether it can be updated using a Refresh Token? I'm not sure that such tasks should be solved through the User Agent.

shawnhankim replied:
- When we get `401` from `/userinfo` or customer's endpoints because access token isn't valid, User Agent can call `/login` to relogin or refresh token based on the status of IdP's token expiration.
- We need to separately discuss this item which I wanted to talk to you. I will show you the use case later.
- Hence, I will remove this comment to avoid confusion as for now.

route443 commented on openid_connect.server_conf:
```diff
@@ -66,6 +69,70 @@
+    # - Browser to show the signed-in user information.
+    # - https://openid.net/specs/openid-connect-core-1_0.html#UserInfo
+    #
+    location = /userinfo {
```
We cannot say that we support the Userinfo endpoint until such time as:
1. These requests must be made through the back channel communication (client/RP/nginx <-> Userinfo).
2. We need to parse JSON response.
3. Perform response validation in accordance with the requirement.
4. Save keys and values in a keyval.
At the same time, we are not required to validate the Access Token to perform this request.

shawnhankim replied:
- Yeah, I originally thought and tried your approach.
- But, the requirement of NMS-ACM was to proxy and return the IdP's response itself to the SPA back then.
- Yeah, it is good point of access token validation too. If we validate it using local JWK file in the backend, and if it isn't invalid, then it reduces latency. However, if it is valid, then the IdP validates it using Bearer token again. 
- I will try achieving both requirements between NGINX Plus and NMS-ACM. Afterwards, you can review, and I will revise more things.  Thanks!

shawnhankim replied:
- For the use case where SPA wants to show user profile in detail like `www.example.com "sub,name,preferred_username,given_name,family_name,email,photo"`, I have extracted only configured claims in `$oidc_userinfo_response_data` and return the JSON rather than directly returning all claims from the OP.
- Let me know if you have other thoughts.

route443 reviewed (changes requested):
As you can see, we are not using Conventional Commits for this project. So I wouldn't use this approach unnecessarily.

route443 commented on frontend.conf:
```diff
@@ -33,4 +46,106 @@ server {
+    #      OneLogin, Ping Identity, etc as for now.
+    #
+    location /v1/api/example {
+        auth_jwt "" token=$access_token;      # Use $session_jwt for Azure AD
```
I also forgot to note. That if we didn't find an Access Token in the tokenset response from the AS? We must properly handle this case, and only in cases where it is necessary.

shawnhankim commented:
Thanks @route443 for your time and great review. 👍 I will address them next week.

shawnhankim commented on frontend.conf:
```diff
@@ -33,4 +46,106 @@ server {
+# -----------------------------------------------------------------------------#
+
+#
+# Upstream server for proxing to the frontend site.
```
- Yes, you are right! 
- The purpose is for customers to easily follow the example of demo which we are going to create the feature branch (e.g, `demo`) instead of the 3rd-party link as discussed. 
- Given the kind of demo, customers can quickly set up IdP, configure NGINX Plus, and develop their application using NGINX Plus OIDC.
- Therefore, I just added the link as the `demo` branch is not created yet.
- However, I will remove the link as for now. 
- Let me know once you create a feature branch such as `demo` so that I can add it to there.

shawnhankim commented on frontend.conf:
```diff
@@ -33,4 +46,106 @@ server {
+server {
+    # Enable when debugging is needed.
+    error_log  /var/log/nginx/error.log  debug; # Reduce severity level as required
+    access_log /var/log/nginx/access.log main;
```
Thanks for finding it.

shawnhankim commented on frontend.conf:
```diff
@@ -33,4 +46,106 @@ server {
+    server_name nginx.oidc.test;
+
+    # Replace the following files with your certificate.
+    ssl_certificate     /etc/ssl/nginx/nginx-repo.crt;
```
Thanks for your opinion. Per the request of one example, I will remove this to just guide to use SSL/TLS in production using the existing comment.

shawnhankim commented on frontend.conf:
```diff
@@ -33,4 +46,106 @@ server {
+    #
+    #  - To ensure that client requests access the API securely, access token is
+    #    used for API authorization.
+    #    + Most of IdP generate an access token for API authorization of IdP's
```
I agree with you. Thanks!

shawnhankim commented on frontend.conf:
```diff
@@ -33,4 +46,106 @@ server {
+    #  - The site is protected with OpenID Connect(OIDC) by calling the API 
+    #    endpoint of `/login` when users click 'login' button.
+    #
+    location / {
```
- Thanks for your opinion.
- Per the requirement of NMS-ACM Developer Portal:
   - The landing page just show default contents should not be protected by OIDC using public APIs before login. 
   - After logged-in, there are more information that can be shown using partner/private APIs that are protected.
- There are also many services that need to be shown as a landing page from the `/` location that is not protected by OIDC which means before login.
- In case of this, `/` and public APIs are not required to be protected, and private/public APIs need to be protected.
- I will revise the example and show it to you once I am ready. In the meantime, feel free to let me know if you have any better idea or comment.

shawnhankim replied:
Revised the example.

shawnhankim commented on frontend.conf:
```diff
@@ -33,4 +46,106 @@ server {
+    #
+    # Backend API example to interact with proxied backend service:
+    #
+    #  - This API resource is protected by access token which is received by IdP
```
Yes, the access token is obtained from the AS. I will revise the comment.

shawnhankim commented on frontend.conf:
```diff
@@ -33,4 +46,106 @@ server {
+    #      endpoints (like /userinfo) as well as customer's endpoints.
+    #    + But Azure AD generate two types of access token for API authorization
+    #      of Microsoft graph API endpoints and customers' endpoints.
+    #    + Therefore, we recommend that you use $session_jwt for Azure AD and
```
Sure, I will remove it.

shawnhankim commented on openid_connect.server_conf:
```diff
@@ -66,6 +69,70 @@
+    #
+    # User information endpoint for the following purposes:
+    # - Browser to periodically check if your are signed-in based on status code.
+    # - Browser to show the signed-in user information.
```
**Use Case:**
- It is used when customer system needs to show user profile (e.g., photo, last name, first name, email, address, etc.) by requesting to IdP. So customers can show them and don't need to store PII into their system.
- The NMS-ACM Dev Portal uses `prefered_username` and `name`, and it is going to use `email` later.

**What if I don't want this information to be exposed to the end user?**
- The infrastructure owner of customer is responsible for and can configure the fields when setting up user pool in the IdP.
- So, it should be configurable not only in NGINX conf but also in IdP by the customer.

shawnhankim replied:
The `$oidc_userinfo_response_data` is empty as default value, then the information is not exposed for the scenario if customers don't want this info to be exposed to the end user.

route443 commented on openid_connect.server_conf:
```diff
@@ -66,6 +69,70 @@
+    # Login endpoint to start OIDC flow when a user clicks 'login' button in the
+    # landing page.
+    #
+    location = /login {
```
Seems like a specific configuration example rather than a OIDC functional location. I think that we should not expose any additional locations by default unless it is absolutely necessary from the functional point of view.

route443 commented on openid_connect.server_conf:
```diff
@@ -66,6 +69,70 @@
+    }
+
+    #
+    # V2 Logout: The following features are added in the NGINX R29.
```
There is nothing specific to NGINX Plus R29 here, I think we should not mislead anyone.

route443 commented on openid_connect.server_conf:
```diff
@@ -66,6 +69,70 @@
+    # - Sample logout page for your OIDC simulation.
+    # - TODO: Custom logout parameters will be separately added.
+    #
+    location = /v2/logout {
```
As I mentioned earlier, we should merge this location with `/logout`. In other words, the location for initializing the logout process must be unified.

route443 commented on openid_connect.server_conf:
```diff
@@ -66,6 +69,70 @@
+        js_content oidc.v2logout;
+    }
+
+    location = /v2/_logout {
```
The same is true for `/v2/_logout` as well. We need to think about unification with `/_logout`.

route443 commented on openid_connect.server_conf:
```diff
@@ -66,6 +69,70 @@
+    }
+
+    #
+    # V1 Logout (NGINX R28): 
```
It's not worth mentioning the NGINX Plus version here either.

route443 commented on openid_connect_configuration.conf:
```diff
@@ -35,7 +43,54 @@ map $host $oidc_scopes {
+    # default https://www.nginx.com;
+}
+
+map $host $return_token_to_client_on_login {
```
Since the `ID Token` is stored in a variable, it can also be returned from the app-specific config. Not sure if we need this option.

shawnhankim commented:
@route443 
- Thanks for your review in detail.
- Addressed most of comments, and added some context.
- Feel free to let me know if you need more changes, or if I miss anything.
- FYI: I will squash the commits after your review.

shawnhankim commented:
@route443 
FYI: If you want to test the changes, you could either find the following link for your convenience or use your test environment:
- https://github.com/nginx-openid-connect/nginx-oidc-examples/tree/main/001-oidc-local-test
